pkg/docker/docker.go

```diff
@@ -0,0 +1,277 @@
+	RunAs string
+
+	// RootDir is where P2 writes files related to this container, not to be
+	// confused with the container filesystem's location on disk which is managed
+	// by dockerd
+	RootDir string
+
+	// RestartTimeout is poorly named in the context of docker containers, since
+	// we never restart a container, only stop them and start them. In this case
+	// it's the amount of time we ask docker to give a container to stop after
+	// SIGTERM is sent before SIGKILL
+	RestartTimeout time.Duration
+
+	// Unique name for log messages to uniquely identify the pod and container
+	ServiceID_ string
+
+	// DockerClient is the client used to start and stop containers
+	DockerClient *dockerclient.Client
+
+	// EntryPoints are the commands to run in the container
+	EntryPoints []string
+
+	// ParentCgroupID is the name of the pod-wide cgroup. Docker containers will
+	// be configured with cgroups inheriting from this one.
+	ParentCgroupID string
+
+	// CPUQuota is the number of CPUs the cgroup is allowed to use. Technically, this
+	// measurement is in "millions of microseconds per second", so it should be multiplied
+	// by a million before being passed to docker
+	CPUQuota int
+
+	// CgroupMemorySize is the max allowed memory docker should set for the cgroup for
+	// the container
+	CgroupMemorySize size.ByteCount
+
+	// RestartPolicy_ determines what behavior docker should have when the pod exits.
+	// Typically it will indicate that docker should always restart the container
+	// for long-running services (e.g. servers)
+	RestartPolicy_ runit.RestartPolicy
+}
+
+func (l *Launchable) Disable() error {
+	// TODO: implement
+	return nil
+}
+
+func (l *Launchable) EnvDir() string {
+	return filepath.Join(l.RootDir, "env")
+}
+
+func (l *Launchable) EnvVars() map[string]string {
+	// TODO: implement
+	return nil
+}
+
+func (l *Launchable) Executables(_ *runit.ServiceBuilder) ([]launch.Executable, error) {
+	// We don't make use of runit for docker launchables, so there are no
+	// executables to supply. Instead we'll just send "docker run" and
+	// "docker stop" commands to dockerd.
+	return nil, nil
+}
+
+func (l *Launchable) GetRestartTimeout() time.Duration {
+	return l.RestartTimeout
+}
+
+func (l *Launchable) ID() launch.LaunchableID {
+	return l.LaunchableID
+}
+
+func (l *Launchable) InstallDir() string {
+	// There is no install directory for a docker container
+	return ""
+}
+
+func (l *Launchable) Installed() bool {
+	// docker pull is idempotent and cheap, so don't worry about this
+	return false
+}
+
+func (l *Launchable) Launch(_ *runit.ServiceBuilder, _ runit.SV) error {
+	if l.DockerClient == nil {
+		return util.Errorf("docker client was not initialized, can't launch docker launchable")
+	}
+
+	ctx := context.TODO()
+	envVars := []string{}
+	envFiles, err := ioutil.ReadDir(l.EnvDir())
+	if err != nil {
+		return util.Errorf("could not read from %s: %v", l.EnvDir(), err)
+	}
+	for _, file := range envFiles {
```

Comment:
is there any code in p2-exec that can be reused here? it probably knows how to turn a directory into a map[string]string or could be trivially modified to do so

Reply:
The code in p2-exec in func loadEnvDir seems to be similar code except it does some validation checks. These checks include ensuring there's no equal sign in the filename, taking only the first line of data in the file, discarding trailing white space, and replace null characters with newline. Would we want that validation here as well?